File: src/DepthCameraBringup.cpp
#include <chrono>
#include <memory>
#include <string>
#include "ros/ros.h"
#include <sensor_msgs/msg/point_cloud.h>

#include <std_msgs/msg/float32_multi_array.h>
#include <vector>

#include "ArducamTOFCamera.hpp"

using namespace std::chrono_literals;
using namespace Arducam;

ArducamTOFCamera tof;

class TOFPublisher : public ros::Node
{
public:
    TOFPublisher() : Node("arducam"), pointsize_(43200)
    {   
        ros::init(argc, argv, "depth_camera");
        pc2_msg_ = std::make_shared<sensor_msgs::msg::PointCloud>();
        pc2_msg_->points.resize(pointsize_);
        pc2_msg_->channels.resize(1);
        pc2_msg_->channels[0].name = "intensities";
        pc2_msg_->channels[0].values.resize(pointsize_);
        depth_msg_ = std::make_shared<std_msgs::msg::Float32MultiArray>();
        depth_msg_->layout.dim.resize(2);
        depth_msg_->layout.dim[0].label = "height";
        depth_msg_->layout.dim[0].size = 180;
        depth_msg_->layout.dim[0].stride = 43200;
        depth_msg_->layout.dim[1].label = "width";
        depth_msg_->layout.dim[1].size = 240;
        depth_msg_->layout.dim[1].stride = 240;
        
        publisher_ = nh->advertise<sensor_msgs::msg::PointCloud>("point_cloud", 10);
        publisher_depth_ = nh->advertise<std_msgs::msg::Float32MultiArray>("depth_frame", 10);

        timer_ = nh.createTimer(
            ros::Duration(0.05), std::bind(&TOFPublisher::update, this));
    }

private:
    void generateSensorPointCloud()
    {
        ArducamFrameBuffer *frame;
        do
        {
            frame = tof.requestFrame(200);
        } while (frame == nullptr);
        depth_frame.clear();
        float *depth_ptr = (float *)frame->getData(FrameType::DEPTH_FRAME);
        float *amplitude_ptr = (float *)frame->getData(FrameType::AMPLITUDE_FRAME);
        unsigned long int pos = 0;
        for (int row_idx = 0; row_idx < 180; row_idx++)
            for (int col_idx = 0; col_idx < 240; col_idx++, pos++)
            {
                if (amplitude_ptr[pos] > 30)
                {
                    float zz = depth_ptr[pos]; 
                    pc2_msg_->points[pos].x = (((120 - col_idx)) / fx) * zz;
                    pc2_msg_->points[pos].y = ((90 - row_idx) / fy) * zz;
                    pc2_msg_->points[pos].z = zz;
                    pc2_msg_->channels[0].values[pos] = depth_ptr[pos];
                    depth_frame.push_back(depth_ptr[pos]);
                }
                else
                {
                    pc2_msg_->points[pos].x = 0;
                    pc2_msg_->points[pos].y = 0;
                    pc2_msg_->points[pos].z = 0;
                    pc2_msg_->channels[0].values[pos] = 0;
                    depth_frame.push_back(0);
                }
            }
        tof.releaseFrame(frame);
        pc2_msg_->header.frame_id = frame_id_;

        depth_msg_->data = depth_frame;
    }
    void update()
    {
        generateSensorPointCloud();
        pc2_msg_->header.stamp = now();
        publisher_->publish(*pc2_msg_);
        publisher_depth_->publish(*depth_msg_);
    }

    std::string frame_id_ = "sensor_frame";
    std::vector<float> depth_frame;
    sensor_msgs::msg::PointCloud::SharedPtr pc2_msg_;
    std_msgs::msg::Float32MultiArray::SharedPtr depth_msg_;

    size_t pointsize_;
    ros::Timer::SharedPtr timer_;
    ros::Publisher<sensor_msgs::msg::PointCloud>::SharedPtr publisher_;
    ros::Publisher<std_msgs::msg::Float32MultiArray>::SharedPtr publisher_depth_;
    ros::NodeHandle::SharedPtr nh;
    float fx = 240 / (2 * tan(0.5 * M_PI * 64.3 / 180));
    float fy = 180 / (2 * tan(0.5 * M_PI * 50.4 / 180));
};

int main(int argc, char *argv[])
{
    
    if (tof.init(Connection::CSI))
    {
        printf("initialize fail\n");
        exit(-1);
    }

    if (tof.start())
    {
        printf("start fail\n");
        exit(-1);
    }
    tof.setControl(ControlID::RANGE,4);

    printf("pointcloud publisher start\n");

    setvbuf(stdout, NULL, _IONBF, BUFSIZ);
    ros::spin(std::make_shared<TOFPublisher>());
    ros::shutdown();
    return 0;
}

File: src/DepthCameraBringup.py
#!/usr/bin/env python3

import rospy
from sensor_msgs.msg import CompressedImage, Image, CameraInfo, PointCloud2, PointField, PointCloud, ChannelFloat32
from geometry_msgs.msg import Point32
from utils.bridge import numpy_to_compressed_imgmsg
from cv_bridge import CvBridge
import sys
import cv2
import numpy as np
import ArducamDepthCamera as ac
import math


class PointCloudComputer:

    def __init__(self, nrows: int, ncols: int, vertical_fov: float, horizontal_fov: float):
        self.nrows = nrows
        self.ncols = ncols
        self.fy = nrows/(2*np.tan(0.5*math.pi*vertical_fov/nrows))
        self.fx = ncols/(2*np.tan(0.5*math.pi*horizontal_fov/nrows))

        self.row_arr = np.tile(np.arange(self.nrows), (self.ncols, 1)).T
        self.col_arr = np.tile(np.arange(self.ncols), (self.nrows, 1))
        self.x_multiplier = ((self.col_arr*-1)+(self.ncols/2))/self.fx
        self.y_multiplier = ((self.row_arr*-1)+(self.nrows/2))/self.fy

        self.zero = np.zeros((nrows, ncols))
        self.msg = PointCloud()
        # self.msg.height=1
        # self.msg.width=ncols*nrows
        # self.msg.fields = [PointField()]*3
        # self.msg.fields[0].name = "x"
        # self.msg.fields[1].name = "y"
        # self.msg.fields[2].name = "z"
        # for i in range(3):
        #     self.msg.fields[i].offset = i*(nrows*ncols)
        #     self.msg.fields[i].datatype=7
        #     self.msg.fields[i].count = nrows*ncols
        self.msg.points = [Point32()]*(self.nrows*self.ncols)
        self.msg.channels = [ChannelFloat32()]
        self.msg.channels[0].values = [0]*(self.nrows*self.ncols)
        self.msg.channels[0].name = "intensities"


    def camera_info_msg(self):
        cinfo = CameraInfo()
        cinfo.height = self.nrows
        cinfo.width = self.ncols
        # cinfo.distortion_model=[self.fx, 0, 0, 0, self.ncols/2]
        cinfo.K = [self.fx, 0, self.ncols/2, 0, self.fy, self.nrows/2, 0, 0, 1]
        cinfo.P = [self.fx, 0, self.ncols/2, 0, 0, self.fy, self.nrows/2, 0, 0, 0, 1, 0]
        cinfo.header.stamp = rospy.Time().now()
        return cinfo




    '''
    Takes in the depth and amplitude data from the depth camera and uses it to compute
    a pointcloud relative to the camera. Returns the PointCloud ROS message as output
    '''
    def numpy_to_pcmsg(self, depth: np.ndarray, amplitude: np.ndarray) -> PointCloud2:
        rospy.loginfo(str(np.max(depth)))
        rospy.loginfo(str(np.min(depth))) 
        # z_arr = np.where(amplitude > 30, depth, self.zero)
        # x_arr = np.where(amplitude > 30, ((self.col_arr - (self.ncols/2)) / self.fx) * z_arr, self.zero)
        # y_arr = np.where(amplitude > 30, ((self.row_arr - (self.nrows/2)) / self.fy) * z_arr, self.zero)
        # z_arr = depth
        # x_arr = np.multiply(self.x_multiplier, depth)
        # y_arr = np.multiply(self.y_multiplier,depth)
        # self.msg.data = list(x_arr.flatten()) + list(y_arr.flatten()) + list(z_arr.flatten())
        count = 0
        self.msg.channels[0].values = list(depth.flatten())
        for row_idx in range(self.nrows):
            for col_idx in range(self.ncols):
                if amplitude[row_idx, col_idx] > 30:
                    self.msg.points[count].x = ((120-col_idx)/self.fx)*depth[row_idx, col_idx]
                    self.msg.points[count].y = ((90-row_idx)/self.fy)*depth[row_idx, col_idx]
                    self.msg.points[count].z = depth[row_idx, col_idx]
                else:
                    self.msg.points[count].x = 0
                    self.msg.points[count].y = 0
                    self.msg.points[count].z = 0
                count = count + 1
        # rospy.loginfo(np.all(x_arr == x_arr[0,0]))
        self.msg.header.stamp = rospy.Time.now()
        self.msg.header.frame_id = "pointcloud"
        return self.msg





MAX_DEPTH = 4




'''
Used to process the amplitude and depth data from the depth camera into an
image in the form of a numpy array.
'''
def process_frame(depth_buf: np.ndarray, amplitude_buf: np.ndarray) -> np.ndarray:
        
    depth_buf = np.nan_to_num(depth_buf)

    amplitude_buf[amplitude_buf<=7] = 0
    amplitude_buf[amplitude_buf>7] = 255

    depth_buf = (1 - (depth_buf/MAX_DEPTH)) * 255
    depth_buf = np.clip(depth_buf, 0, 255)
    result_frame = depth_buf.astype(np.uint8)  & amplitude_buf.astype(np.uint8)
    return result_frame


if __name__ == "__main__":
    rospy.init_node("pupper_depth_camera")
    pub = rospy.Publisher("camera/depth/image_raw", Image, queue_size=1)
    info_pub = rospy.Publisher("camera/depth/camera_info", CameraInfo, queue_size=1)
    point_cloud_pub = rospy.Publisher("/points", PointCloud, queue_size=1)
    bridge = CvBridge()
    dev = rospy.get_param("dev")
    cam = ac.ArducamCamera()
    rate = rospy.Rate(30)

    calc = PointCloudComputer(180, 240, 50.4, 64.3)
    
        
    if cam.init(ac.TOFConnect.CSI,dev) != 0 :
        rospy.logerr("initialization failed")
    if cam.start(ac.TOFOutput.DEPTH) != 0 :
        rospy.logerr("Failed to start camera")
    
    cam.setControl(ac.TOFControl.RANG,MAX_DEPTH)
    while not rospy.is_shutdown():
        frame = cam.requestFrame(200)
        rate.sleep()
        if frame != None:
            depth_buf = frame.getDepthData()
            amplitude_buf = frame.getAmplitudeData()
            cam.releaseFrame(frame)
            calc.numpy_to_pcmsg(depth_buf, amplitude_buf)
            point_cloud_pub.publish(calc.msg)
            amplitude_buf*=(255/1024)
            amplitude_buf = np.clip(amplitude_buf, 0, 255)
            # rospy.loginfo(str(depth_buf.dtype))
            # info = calc.camera_info_msg()
            img = process_frame(depth_buf,amplitude_buf)
            # rospy.loginfo(img.shape)
            # info_pub.publish(info)
            pub.publish(bridge.cv2_to_imgmsg(img, encoding="mono8"))
            
        else:
            rospy.logwarn("Did not recieve frame")
    # finally:
    #     cam.stop()
